crates/sui-indexer/src/handlers/checkpoint_handler.rs

```diff
@@ -188,6 +231,62 @@ where
         }
     }
 
+    async fn start_checkpoint_commit(&self) -> Result<(), IndexerError> {
+        info!("Indexer checkpoint commit task started...");
+        loop {
+            let mut checkpoint_receiver_guard = self.checkpoint_receiver.lock().await;
+            let indexed_checkpoint = checkpoint_receiver_guard.recv().await;
```

Comment:
hmm, where do we guarantee that checkpoint is processed in order by sequence number?

Reply:
Previously I found this to be a good template to use

```
#[tokio::main]
async fn main() -> Result<(), Error> {
    let (tx, rx): (UnboundedSender<(usize, Vec<u8>)>, UnboundedReceiver<(usize, Vec<u8>)>) = unbounded_channel();
    let next_to_process = Arc::new(AtomicUsize::new(0));

    let max_concurrent_downloads = 5;
    let download_semaphore = Arc::new(Semaphore::new(max_concurrent_downloads));

    // Use reqwest::Client with connection pooling
    let client = reqwest::Client::new();

    let download_task = task::spawn(async move {
        let mut index = 0;
        loop {
            let url = format!("https://example.com/file{}.bin", index);
            let permit = download_semaphore.acquire().await;
            let tx = tx.clone();
            let download_semaphore = download_semaphore.clone();
            let client = client.clone();

            task::spawn(async move {
                match client.get(&url).await?.bytes().await {
                    Ok(bytes) => {
                        tx.send((index, bytes.to_vec())).unwrap();
                    }
                    Err(e) => {
                        eprintln!("Error downloading file {}: {:?}", index, e);
                    }
                }
                drop(permit);
                Ok::<_, Error>(())
            });

            index += 1;
        }
    });

    let next_to_process_clone = next_to_process.clone();
    let process_task = task::spawn(async move {
        let mut buffer: Vec<Option<Vec<u8>>> = Vec::new();
        for (index, data) in rx.iter() {
            if index >= buffer.len() {
                buffer.resize(index + 1, None);
            }
            buffer[index] = Some(data);

            while let Some(data) = buffer[next_to_process_clone.load(Ordering::Relaxed)].take() {
                process_file(next_to_process_clone.load(Ordering::Relaxed), &data);
                next_to_process_clone.fetch_add(1, Ordering::Relaxed);
            }
        }
    });

    let _ = tokio::try_join!(download_task, process_task);

    Ok(())
}
```

Ordering is guaranteed by `next_to_process`. It also has buffering built-in as well

Reply:
> where do we guarantee that checkpoint is processed in order by sequence number?

mspc [channel](https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html) is a basically a `VecDeque` with `buffer_size` for multi threads. more specifically
> All data sent on the [Sender](https://doc.rust-lang.org/std/sync/mpsc/struct.Sender.html) will become available on the [Receiver](https://doc.rust-lang.org/std/sync/mpsc/struct.Receiver.html) in the same order as it was sent


the template you shared is very similar semantically, except the fact that it uses `unbounded_channel` instead of `channel`.

Reply:
Got it. There's one major difference though, your code is downloading each checkpoint sequentially while the code snippet parallelizes the downloading of multiple checkpoints. The latter could be useful when network latency is high.

Reply:
good point, we can do that as an incremental change if needed, if download turns out to be super slow, the data freshness will suffer even with parallel downloading, so we might still want to optimize the downloading latency even not only for throughput.